File: pt5/1191.k-concatenation-maximum-sum.java
import java.util.*;

class Solution {
    public int kConcatenationMaxSum(int[] arr, int k) {
        long mod = 1_000_000_007L;

        if (k == 1) {
            return (int) (kadane(arr, 1) % mod);
        }

        long best = kadane(arr, 2);
        long total = 0;
        for (int value : arr) {
            total += value;
        }
        if (total > 0) {
            best += (long) (k - 2) * total;
        }

        return (int) (best % mod);
    }

    private long kadane(int[] arr, int times) {
        long best = 0;
        long current = 0;
        for (int t = 0; t < times; t++) {
            for (int value : arr) {
                current = Math.max(0, current + value);
                best = Math.max(best, current);
            }
        }
        return best;
    }
}

File: pt5/1191.k-concatenation-maximum-sum.go
package main

func kConcatenationMaxSum(arr []int, k int) int {
	const mod int64 = 1_000_000_007

	kadane := func(times int) int64 {
		best, current := int64(0), int64(0)
		for t := 0; t < times; t++ {
			for _, value := range arr {
				current += int64(value)
				if current < 0 {
					current = 0
				}
				if current > best {
					best = current
				}
			}
		}
		return best
	}

	if k == 1 {
		return int(kadane(1) % mod)
	}

	best := kadane(2)
	total := int64(0)
	for _, value := range arr {
		total += int64(value)
	}
	if total > 0 {
		best += int64(k-2) * total
	}
	return int(best % mod)
}
